### ISSUE TITLE
NTLM with IIS Express

### ISSUE THREAD

**Issue by bjowes** [bjowes/cypress-ntlm-auth#21]
IIS Express and Windows Authentication does not always work as expected. 
Given that you are:
* running Visual Studio as a regular user (not as Administrator)
* debugging you website from Visual Studio
* have enabled Windows Authentication for your project 
* have disabled Anonymous Authentication for your project
* hosting on IIS Express during debugging
my observation is that it works fine as long as you are using the same account when signing on to the site as when logging on to the computer with. When you try to use another account, it will mostly fail with a HTTP status code 500.

The reliable solution to fix this is to run IIS Express as Administrator (meaning you have to run Visual Studio as Administrator). Not a great solution, but Microsoft hasn't provided any other options. I have tried to get this verified in the documentation, but I cannot find anyting on the topic of requirements for IIS Express to run Windows Authentication. There is an article about running [IIS Express without Administrator priveliges](https://docs.microsoft.com/en-us/iis/extensions/using-iis-express/running-iis-express-without-administrative-privileges) but it doesn't mention authentication at all.

My own speculation is that true Windows Authentication will partly impersonate the user, and this requires Administrator priveliges or UAC popup - but the UAC popup is disabled when running IIS Express.  

To sum up, this is not an issue with this plugin but with the Authentication model i IIS Express. If I find any better ways to debug sites properly from Visual Studio I will add that info.

**Comment by svopex** [comment]
@bjowes: 
"_..my observation is that it works fine as long as you are using the same account when signing on to the site as when logging on to the computer with..._"
It's not working for me. But I use windows DOMAIN user. The resources failed me with error 500, normal pages works ok.

**Comment by bjowes** [comment]
Yes @svopex , during further testing I found this to be a bit random. I am running a web api that makes web requests to another site (which also requires authentication). This works sometimes when using my own user, but doesn't work when signing in as another user.

**Comment by SacuL** [comment]
Hi @bjowes ,

Thank you for this plugin.
I'm also facing this IIS express problem, even when using the same user logged into the running machine.
As you suggested, it works if I run VS as Admin.

**Comment by bjowes** [comment]
Hi @SacuL - just did some more research on this but I cannot find any other alternatives than running as admin. 

I did find a note though that [loopback check](https://support.microsoft.com/en-us/help/926642/error-message-when-you-try-to-access-a-server-locally-by-using-its-fqd) might be enabled on Windows 10, meaning that if you are accessing an IIS Express site by a "real" FQDN (not localhost) you may need to add some registry keys to allow loopback authentication. But you would still need to run as admin.

**Comment by Ciruwin** [comment]
Hi @bjowes

I'm a newcomer to cypress. I'm trying to achieve the same thing but not successful sofar.
I've started VS as administrator, but cypress-ntlm can still not connect to the web site. I get the following error:

_"cy.visit() failed trying to load: http://localhost:50000/ We attempted to make an http request to this URL but the request failed without a response. We received this error at the network level: > Error: socket hang up"_

I wonder whether the issue could be somehow related to my proxy configurations.
When I start cypress then it shows nothing under proxy configuration.
When I start cypress-ntlm it shows the following proxy configuration:
Proxy Server: http://127.0.0.1:52532
Proxy Bypass List: <-loopback>
Is this proxy configuration correct for trying to access  http://localhost:50000/ on IIS Express?
Is there anything else I could check to see why I can't access http://localhost:50000/ on IIS Express?

**Comment by bjowes** [comment]
Hi @Ciruwin - when you get network level errors, that must be a different issue. Unable to authenticate with IIS Express should result in a response with code 401 or 500.

As for the proxy settings, cypress-ntlm will set these for you when it is activated. This is required since all traffic from cypress must also pass through the plugin to handle the authentication handshake.

Since you have a network error, I would start with a simpler setup. Run just cypress (without the plugin) and try to get that working first. If you can reach the site which requires authentication, you should get a 401. First place to look is probably if your site is accepting http or only https.

**Comment by Ciruwin** [comment]
Hi @bjowes
I've tried the simpler setup as you suggestet. If I run cypress alone I can visit the site and execute further test steps successfully.
The site accepts http. Cypress without the plugin can reach the site without problems.
I tried an even simpler setup with the following steps:

- I created a new ASP.net Forms Webapplication (out of the box) with Visual Studio 2019 (By the way my System is Windows 10).
- I installed cypress and cpyress-ntml locally as follows:
     - navigated into the Webapplication folder
     - executed:
          - npm init
          - npm install cypress --save-dev
          - npm install cypress-ntlm-auth --save-dev
          - npx cypress open  (so that cypress folder and cypress.json are created)
- Started the new Webapplication as administrator
- Added the following test script to ..\ Webapplication\cypress\integration\test.js:

describe('My First Cypress Test', function () {
    it('Does not do much meaningful',
        function() {
            cy.visit('https://localhost:44368/');
        });
    })

-	Navigated into the Webapplication folder
-	Openen cypress with: npx cypress open
-	Executed the test created above --> The website could be reached successfully.
-	Ended cypress
-	Opened cypress-ntml with: npx cypress-ntlm open
-	Executed the test created above --> Received the following error regardless of any browser I selected in cypress:

_cy.visit() failed trying to load: https://localhost:44368/ We attempted to make an http request to this URL but the request failed without a response. We received this error at the network level: > Error: Error establishing proxy connection. Response from server was: HTTP/1.1 502 Bad Gateway_

Is there anything else I could try to do?

**Comment by bjowes** [comment]
Well that is strange. Would you be able to share your simple test application, for instance as a public github repo? 
Then I can try reproducing it locally.

**Comment by Ciruwin** [comment]
Hi @bjowes

Here is my test project:

[https://github.com/Ciruwin/WebApplication1](url)

The project includes the following cypress test script:

describe('My First Cypress Test', function () {
    it('Does not do much meaningful',
    function() {
    cy.visit('https://localhost:44368/');
    });
})

If I run that script with the cypress configured in this project then the server does not respond as described in my previous post. If I change the same script and replace that local URL with the URL of a remote IIS Server and run the script with the same cypress configuration from the same project then the server responds.

**Comment by bjowes** [comment]
I've tried it out now, and it works on my machine. I added `cy.ntlmSso(['localhost'])` before the cy.visit command and cypress could then authenticate and show the page properly (when started with `npx cypress-ntlm open`. Hence there must be something different on your machine. 

Which browser are you testing with?
Could you check if any of these environment variables are set on your machine: HTTP_PROXY, HTTPS_PROXY, NO_PROXY

**Comment by Ciruwin** [comment]
I tried it again with the command cy.ntlmSso(['localhost']) as you mentioned above. I still get the same error for cy.visit.
I've tried with all four browsers which cypress provides (Chrome 90, Firefox 88, Edge 90 and Electron 89), all lead to the same error.
None of the proxy environment variables you mentioned above is set on my machine.
I've also tried it with my firewall deactivated, to rule out any effects from firewall, but I got the same result.

**Comment by bjowes** [comment]
Ok, then it is time to add some debug logging. Set the environment variable
DEBUG=cypress:plugin:ntlm-auth
before starting cypress-ntlm. The logs should appear in the console window.
Then add the logs to this issue.

**Comment by Ciruwin** [comment]
I don't know whether I'm setting the environment varialbe correctly or not but this is what I've tried:

- I openend VS as Administrator.
- I opened a powershell window from within VS
- I entered the following commands into the powershelll window:
     - PS C:\Temp2\TestProject\WebApplication1> set DEBUG=cypress:plugin:ntlm-auth
     - PS C:\Temp2\TestProject\WebApplication1> npx cypress-ntlm open
- In the cypress windows I click on the test
- The same error as before is reported by cypress and in the powershell window I get the following logs:

GET /__/ 200 10.485 ms - -
GET /__cypress/runner/cypress_runner.css 200 2.198 ms - -
GET /__cypress/runner/cypress_runner.js 200 1.420 ms - -
GET /__cypress/iframes/integration/test.js 200 2.837 ms - 892
GET /__cypress/runner/fonts/fa-solid-900.woff2 200 0.796 ms - 76120
GET /__cypress/tests?p=cypress%5Cintegration%5Ctest.js 200 708.494 ms - -
GET /__cypress/tests?p=cypress%5Csupport%5Cindex.js 200 800.464 ms - -

**Comment by bjowes** [comment]
Ok, no in PowerShell you need a different syntax to set the environment
variable. I've tripped on this myself a few times, should probably add it
to my docs.
Set it like this:
$env:DEBUG='cypress:plugin:ntlm-auth'

**Comment by Ciruwin** [comment]
Ok, I've run it again and this time I've got a lot of logs indeed:

**********************************************************************
** Visual Studio 2019 Developer PowerShell v16.7.7
** Copyright (c) 2020 Microsoft Corporation
**********************************************************************
PS C:\Temp2\TestProject\WebApplication1> $env:DEBUG='cypress:plugin:ntlm-auth'
PS C:\Temp2\TestProject\WebApplication1> npx cypress-ntlm open
  cypress:plugin:ntlm-auth Starting ntlm-proxy... +0ms
  cypress:plugin:ntlm-auth NTLM auth config API listening on port: 51735 +5ms
  cypress:plugin:ntlm-auth NTLM auth proxy listening on port: 51736 +18ms
  cypress:plugin:ntlm-auth Startup done! +0ms
  cypress:plugin:ntlm-auth {
  cypress:plugin:ntlm-auth   configApiUrl: 'http://127.0.0.1:51735',
  cypress:plugin:ntlm-auth   ntlmProxyUrl: 'http://127.0.0.1:51736'
  cypress:plugin:ntlm-auth } +0ms
  cypress:plugin:ntlm-auth Opening Cypress... +2ms

  cypress:plugin:ntlm-auth Tunnel to download.cypress.io:443 +7s
  cypress:plugin:ntlm-auth client closed socket, closing tunnel to  download.cypress.io:443 +293ms
GET /__/ 200 4.394 ms - -
GET /__cypress/runner/cypress_runner.css 200 11.768 ms - -
GET /__cypress/runner/cypress_runner.js 200 223.105 ms - -
GET /__cypress/iframes/integration/test.js 200 3.134 ms - 892
GET /__cypress/runner/fonts/fa-solid-900.woff2 200 2.560 ms - 76120
GET /__cypress/tests?p=cypress%5Cintegration%5Ctest.js 200 1457.807 ms - -
GET /__cypress/tests?p=cypress%5Csupport%5Cindex.js 200 1553.276 ms - -
  cypress:plugin:ntlm-auth Created agent for client 127.0.0.1:51756 to target http://127.0.0.1:51735/ +18s
  cypress:plugin:ntlm-auth Request to config API +0ms
  cypress:plugin:ntlm-auth Created untracked agent for target http://127.0.0.1:51735/ +0ms
  cypress:plugin:ntlm-auth Received valid NTLM SSO config +23ms
  cypress:plugin:ntlm-auth Created agent for client 127.0.0.1:51760 to target https://localhost:44368/ +52ms
  cypress:plugin:ntlm-auth Request to https://localhost:44368/ in registered NTLM Hosts (using SSO) +0ms
  cypress:plugin:ntlm-auth PROXY_TO_SERVER_REQUEST_ERROR on /: Error: connect ECONNREFUSED 127.0.0.1:44368
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
  errno: -4078,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 44368
} Destroying client socket. +2ms
  cypress:plugin:ntlm-auth Removed agent for 127.0.0.1:51760 due to socket.close +2ms
  cypress:plugin:ntlm-auth Created agent for client 127.0.0.1:51763 to target https://localhost:44368/ +10ms
  cypress:plugin:ntlm-auth Request to https://localhost:44368/ in registered NTLM Hosts (using SSO) +1ms
  cypress:plugin:ntlm-auth PROXY_TO_SERVER_REQUEST_ERROR on /: Error: connect ECONNREFUSED 127.0.0.1:44368
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
  errno: -4078,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 44368
} Destroying client socket. +0ms
  cypress:plugin:ntlm-auth Removed agent for 127.0.0.1:51763 due to socket.close +1ms
  cypress:plugin:ntlm-auth Created agent for client 127.0.0.1:51766 to target https://localhost:44368/ +1s
  cypress:plugin:ntlm-auth Request to https://localhost:44368/ in registered NTLM Hosts (using SSO) +0ms
  cypress:plugin:ntlm-auth PROXY_TO_SERVER_REQUEST_ERROR on /: Error: connect ECONNREFUSED 127.0.0.1:44368
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
  errno: -4078,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 44368
} Destroying client socket. +2ms
  cypress:plugin:ntlm-auth Removed agent for 127.0.0.1:51766 due to socket.close +0ms
  cypress:plugin:ntlm-auth Created agent for client 127.0.0.1:51770 to target https://localhost:44368/ +2s
  cypress:plugin:ntlm-auth Request to https://localhost:44368/ in registered NTLM Hosts (using SSO) +1ms
  cypress:plugin:ntlm-auth PROXY_TO_SERVER_REQUEST_ERROR on /: Error: connect ECONNREFUSED 127.0.0.1:44368
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
  errno: -4078,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 44368
} Destroying client socket. +1ms
  cypress:plugin:ntlm-auth Removed agent for 127.0.0.1:51770 due to socket.close +1ms
  cypress:plugin:ntlm-auth Removed agent for 127.0.0.1:51756 due to socket.close +2s
  cypress:plugin:ntlm-auth Created agent for client 127.0.0.1:51773 to target https://localhost:44368/ +112ms
  cypress:plugin:ntlm-auth Request to https://localhost:44368/ in registered NTLM Hosts (using SSO) +0ms
  cypress:plugin:ntlm-auth PROXY_TO_SERVER_REQUEST_ERROR on /: Error: connect ECONNREFUSED 127.0.0.1:44368
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
  errno: -4078,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 44368
} Destroying client socket. +2ms
  cypress:plugin:ntlm-auth Removed agent for 127.0.0.1:51773 due to socket.close +0ms

**Comment by bjowes** [comment]
Thanks for the logs. Now it is clear that the network error appears between the plugin and the server. 
It seems that the web site is not listening to the 127.0.0.1 address. When cypress connects directly without the plugin, it will let the browser resolve "localhost", while the plugin assumes that localhost resolves to 127.0.0.1. Possibly localhost resolves to ::1 (the IPv6 variant) on your machine, and the site is listening only to that. 
Here are some things you can investigate to troubleshoot. https://stackoverflow.com/questions/25652409/localhost-is-working-but-not-127-0-0-1-in-windows-7
Seems that for some users it helped to run this command `netsh http add iplisten 127.0.0.1`

**Comment by Ciruwin** [comment]
@bjowes
Great suggestions, that solved the issue. Before executing the netsh command you mentioned above, browsing https://127.0.0.1: 44368 would lead to a connection error and browsing https://[::1]:44368 would lead to an error 400 "The request hostname is invalid".
After executing the command "netsh http add iplisten 127.0.0.1" you mentioned above, browsing https://127.0.0.1: 44368 would also lead to an error 400.
When I then executed the command "npx cypress-ntlm open" and executed my test case, everything worked like magic. 
Thanks a lot for your help.

**Comment by bjowes** [comment]
Closing as there are still no bullet proof solutions for NTLM with IIS Express to document.
